From 7f0483fec8d8fe12a8cb44f203f697f877e39b1a Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:08:18 +0200
Subject: [PATCH] parser_bison: fix length check for ifname in
 ifname_expr_alloc()

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit 122dce6b35205a3df419a5cae9acfd6e83e8725a

commit 122dce6b35205a3df419a5cae9acfd6e83e8725a
Author: Thomas Haller <thaller@redhat.com>
Date:   Mon Oct 23 19:00:47 2023 +0200

    parser_bison: fix length check for ifname in ifname_expr_alloc()

    IFNAMSIZ is 16, and the allowed byte length of the name is one less than
    that. Fix the length check and adjust a test for covering the longest
    allowed interface name.

    This is obviously a change in behavior, because previously interface
    names with length 16 were accepted and were silently truncated along the
    way. Now they are rejected as invalid.

    Fixes: fa52bc225806 ("parser: reject zero-length interface names")
    Signed-off-by: Thomas Haller <thaller@redhat.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 src/parser_bison.y                                | 3 ++-
 tests/shell/testcases/chains/0042chain_variable_0 | 7 +------
 2 files changed, 3 insertions(+), 7 deletions(-)

diff --git a/src/parser_bison.y b/src/parser_bison.y
index 8eff50c..5c2d14b 100644
--- a/src/parser_bison.y
+++ b/src/parser_bison.y
@@ -16,6 +16,7 @@
 #include <stdio.h>
 #include <inttypes.h>
 #include <syslog.h>
+#include <net/if.h>
 #include <netinet/ip.h>
 #include <netinet/tcp.h>
 #include <netinet/if_ether.h>
@@ -158,7 +159,7 @@ static struct expr *ifname_expr_alloc(const struct location *location,
 		return NULL;
 	}
 
-	if (length > 16) {
+	if (length >= IFNAMSIZ) {
 		xfree(name);
 		erec_queue(error(location, "interface name too long"), queue);
 		return NULL;
diff --git a/tests/shell/testcases/chains/0042chain_variable_0 b/tests/shell/testcases/chains/0042chain_variable_0
index 739dc05..a4b929f 100755
--- a/tests/shell/testcases/chains/0042chain_variable_0
+++ b/tests/shell/testcases/chains/0042chain_variable_0
@@ -26,18 +26,13 @@ table netdev filter2 {
 
 rc=0
 $NFT -f - <<< $EXPECTED || rc=$?
-test "$rc" = 0
+test "$rc" = 1
 cat <<EOF | $DIFF -u <($NFT list ruleset) -
 table netdev filter1 {
 	chain Main_Ingress1 {
 		type filter hook ingress device "lo" priority -500; policy accept;
 	}
 }
-table netdev filter2 {
-	chain Main_Ingress2 {
-		type filter hook ingress devices = { d23456789012345, lo } priority -500; policy accept;
-	}
-}
 EOF
 
 
